packages/protocol/contracts/identity/FederatedAttestations.sol

```diff
@@ -0,0 +1,386 @@
+          for (uint256 k = 0; k < attestationsForIssuer.length; k = k.add(1)) {
+            OwnershipAttestation memory attestation = attestationsForIssuer[k];
+            // (identifier, account, issuer) tuples should be unique
+            if (attestation.account == account && !revokedSigners[attestation.signer]) {
+              identifiers[currIndex] = identifier;
+              currIndex = currIndex.add(1);
+              break;
+            }
+          }
+        } else {
+          break;
+        }
+      }
+    }
+    if (currIndex < maxIdentifiers) {
+      // Allocate and fill properly-sized array
+      bytes32[] memory trimmedIdentifiers = new bytes32[](currIndex);
+      for (uint256 i = 0; i < currIndex; i = i.add(1)) {
+        trimmedIdentifiers[i] = identifiers[i];
+      }
+      return trimmedIdentifiers;
+    } else {
+      return identifiers;
+    }
+  }
+
+  // TODO do we want to restrict permissions, or should anyone
+  // with a valid signature be able to register an attestation?
+  modifier isValidUser(address issuer, address account) {
+    require(
+      msg.sender == account ||
+        msg.sender == issuer ||
+        getAccounts().attestationSignerToAccount(msg.sender) == issuer,
+      "User does not have permission to perform this action"
+    );
+    require(!revokedSigners[msg.sender], "User has been revoked ");
+    _;
+  }
+
+  /**
+   * @notice Validates the given attestation and signature
+   * @param identifier Hash of the identifier to be attested
+   * @param issuer Address of the attestation issuer
+   * @param account Address of the account being mapped to the identifier
+   * @param issuedOn Time at which the issuer issued the attestation in Unix time 
+   * @param signer Address of the signer of the attestation
+   * @param v The recovery id of the incoming ECDSA signature
+   * @param r Output value r of the ECDSA signature
+   * @param s Output value s of the ECDSA signature
+   * @return Whether the signature is valid
+   * @dev Throws if signer is revoked
+   * @dev Throws if signer is not an authorized AttestationSigner of the issuer
+   */
+  function isValidAttestation(
+    bytes32 identifier,
+    address issuer,
+    address account,
+    uint256 issuedOn,
+    address signer,
+    uint8 v,
+    bytes32 r,
+    bytes32 s
+  ) public view returns (bool) {
+    require(!revokedSigners[signer], "Signer has been revoked");
+    require(
+      getAccounts().attestationSignerToAccount(signer) == issuer,
+      "Signer has not been authorized as an AttestationSigner by the issuer"
+    );
+    bytes32 structHash = keccak256(
+      abi.encode(EIP712_VALIDATE_ATTESTATION_TYPEHASH, identifier, issuer, account, issuedOn)
+    );
+    address guessedSigner = Signatures.getSignerOfTypedDataHash(
+      eip712DomainSeparator,
+      structHash,
+      v,
+      r,
+      s
+    );
+    return guessedSigner == signer;
```

Comment:
I find it mildly odd that this function indicates failure in two distinct ways: By reverting if the signer is not authorized or by returning false if the (claimed) signer is authorized, but the signature is invalid. I worry this creates a sharp edge where a developer might see the `require` statements and assume that is how any failure is indicated, but forget to check the return value of this function. If that were to happen, the resulting code would     be unprotected. I would advise on having this return `false` for every failure mode, or revert for every failure mode.

Reply:
good point, will revert for all failure modes instead!